### ISSUE TITLE
Union of array shapes wrongly understood

### ISSUE THREAD

**Issue by BenMorel** [vimeo/psalm#10331]
Consider the following example:

```php
/**
 * @param array{string}|array{value: string} $a
 */
function test(array $a): string
{
    if (isset($a['value'])) {
        return $a['value'];
    }

    return $a[0];
}
```

Psalm erroneously complains:

> ERROR: PossiblyUndefinedArrayOffset - 12:12 - Possibly undefined array key $a[0] on array{0?: string, value?: string}

Reproduce: https://psalm.dev/r/55bc75b438

**Comment by robchett** [comment]
This is a longstanding limitation with disjointed array shapes. The workaround for now is to use objects instead of arrays. See ttps://github.com/vimeo/psalm/issues/10002 as the mega-thread.

Sorry.

**Comment by BenMorel** [comment]
@robchett Thanks for the pointer! Closing this one in favour of #10002.
